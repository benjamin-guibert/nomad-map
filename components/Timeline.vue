<template>
  <div class="timeline pt-4">
    <div class="timeline-content container position-relative">
      <timeline-item
        v-for="(travel, index) in reversedTravels"
        :key="index"
        :travel="travel"
        :orientation="getOrientation(index)"
      />
    </div>
  </div>
</template>

<script>
import _ from 'lodash'

import TimelineItem from './TimelineItem.vue'

export default {
  components: {
    TimelineItem
  },
  props: {
    travels: {
      type: Array,
      required: true
    }
  },
  computed: {
    /**
     * Get reversed travels.
     * @return {Array} The reversed travels.
     */
    reversedTravels() {
      return _.reverse(this.travels)
    }
  },
  methods: {
    /**
     * Get the orientation (left of right) of the travel on the timeline based on its index.
     * @param {Number} index The index of the travel.
     * @return {String} The orientation (left or right)
     */
    getOrientation(index) {
      return index % 2 === 0 ? 'left' : 'right'
    }
  }
}
</script>
