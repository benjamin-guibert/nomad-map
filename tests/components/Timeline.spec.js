import { createLocalVue, mount } from '@vue/test-utils'
import BootstrapVue from 'bootstrap-vue'
import Timeline from '@/components/Timeline.vue'

const localVue = createLocalVue()
localVue.use(BootstrapVue)

describe('Timeline', () => {
  const wrapper = mount(Timeline, {
    localVue,
    propsData: {
      travels: [
        {
          dates: {
            from: new Date(2099, 0, 2)
          },
          country: { id: 'GR', name: 'Greece' },
          city: { name: 'Athens', latitude: 37.966667, longitude: 23.716667 },
          id: 6,
          status: {
            id: 'next',
            label: 'Next destination',
            mapLabel: '(next destination)',
            icon: 'bookmark'
          }
        },
        {
          dates: {
            from: new Date(2099, 0, 1),
            to: new Date(2099, 0, 1)
          },
          country: { id: 'FR', name: 'France' },
          city: { name: 'Paris', latitude: 48.856614, longitude: 2.352222 },
          id: 5,
          status: {
            id: 'next',
            label: 'Next destination',
            mapLabel: '(next destination)',
            icon: 'bookmark'
          }
        },
        {
          dates: {
            from: new Date(2019, 0, 2),
            to: new Date(2098, 11, 31)
          },
          country: { id: 'IS', name: 'Iceland' },
          city: {
            name: 'Reykjavík',
            latitude: 64.126518,
            longitude: -21.817438
          },
          id: 4,
          status: {
            id: 'current',
            label: 'Current destination',
            mapLabel: '(current destination)',
            icon: 'map-marker'
          }
        },
        {
          dates: {
            from: new Date(2019, 0, 1),
            to: new Date(2019, 0, 1)
          },
          country: { id: 'FR', name: 'France' },
          city: { name: 'Paris', latitude: 48.856614, longitude: 2.352222 },
          id: 3,
          status: {
            id: 'visited',
            label: 'Visited',
            mapLabel: '(visited)',
            icon: 'check-circle'
          }
        },
        {
          dates: {
            from: new Date(2018, 6, 1),
            to: new Date(2018, 11, 31)
          },
          country: { id: 'HR', name: 'Croatia' },
          city: { name: 'Zagreb', latitude: 45.815, longitude: 15.9819 },
          id: 2,
          status: {
            id: 'visited',
            label: 'Visited',
            mapLabel: '(visited)',
            icon: 'check-circle'
          }
        },
        {
          dates: {
            from: new Date(2018, 0, 1),
            to: new Date(2018, 5, 30)
          },
          country: { id: 'PL', name: 'Poland' },
          city: { name: 'Warsaw', latitude: 52.2297, longitude: 21.0122 },
          id: 1,
          status: {
            id: 'visited',
            label: 'Visited',
            mapLabel: '(visited)',
            icon: 'check-circle'
          }
        },
        {
          country: { id: 'FR', name: 'France' },
          city: { name: 'Paris', latitude: 48.856614, longitude: 2.352222 },
          id: 0,
          status: {
            id: 'home',
            label: 'Home',
            mapLabel: '(home)',
            icon: 'home'
          }
        }
      ]
    },
    stubs: {
      FontAwesomeIcon: true
    }
  })

  test('is a Vue instance', () => {
    expect(wrapper.isVueInstance()).toBeTruthy()
  })

  test('renders', () => {
    expect(wrapper.html()).toMatchSnapshot()
  })

  test('get reversed travels', () => {
    expect(wrapper.vm.reversedTravels).toEqual([
      {
        country: { id: 'FR', name: 'France' },
        city: { name: 'Paris', latitude: 48.856614, longitude: 2.352222 },
        id: 0,
        status: {
          id: 'home',
          label: 'Home',
          mapLabel: '(home)',
          icon: 'home'
        }
      },
      {
        dates: {
          from: new Date(2018, 0, 1),
          to: new Date(2018, 5, 30)
        },
        country: { id: 'PL', name: 'Poland' },
        city: { name: 'Warsaw', latitude: 52.2297, longitude: 21.0122 },
        id: 1,
        status: {
          id: 'visited',
          label: 'Visited',
          mapLabel: '(visited)',
          icon: 'check-circle'
        }
      },
      {
        dates: {
          from: new Date(2018, 6, 1),
          to: new Date(2018, 11, 31)
        },
        country: { id: 'HR', name: 'Croatia' },
        city: { name: 'Zagreb', latitude: 45.815, longitude: 15.9819 },
        id: 2,
        status: {
          id: 'visited',
          label: 'Visited',
          mapLabel: '(visited)',
          icon: 'check-circle'
        }
      },
      {
        dates: {
          from: new Date(2019, 0, 1),
          to: new Date(2019, 0, 1)
        },
        country: { id: 'FR', name: 'France' },
        city: { name: 'Paris', latitude: 48.856614, longitude: 2.352222 },
        id: 3,
        status: {
          id: 'visited',
          label: 'Visited',
          mapLabel: '(visited)',
          icon: 'check-circle'
        }
      },
      {
        dates: {
          from: new Date(2019, 0, 2),
          to: new Date(2098, 11, 31)
        },
        country: { id: 'IS', name: 'Iceland' },
        city: {
          name: 'Reykjavík',
          latitude: 64.126518,
          longitude: -21.817438
        },
        id: 4,
        status: {
          id: 'current',
          label: 'Current destination',
          mapLabel: '(current destination)',
          icon: 'map-marker'
        }
      },
      {
        dates: {
          from: new Date(2099, 0, 1),
          to: new Date(2099, 0, 1)
        },
        country: { id: 'FR', name: 'France' },
        city: { name: 'Paris', latitude: 48.856614, longitude: 2.352222 },
        id: 5,
        status: {
          id: 'next',
          label: 'Next destination',
          mapLabel: '(next destination)',
          icon: 'bookmark'
        }
      },
      {
        dates: {
          from: new Date(2099, 0, 2)
        },
        country: { id: 'GR', name: 'Greece' },
        city: { name: 'Athens', latitude: 37.966667, longitude: 23.716667 },
        id: 6,
        status: {
          id: 'next',
          label: 'Next destination',
          mapLabel: '(next destination)',
          icon: 'bookmark'
        }
      }
    ])
  })
})
