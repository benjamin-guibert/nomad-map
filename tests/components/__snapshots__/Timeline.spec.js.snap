// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Timeline renders 1`] = `
"<div class=\\"timeline pt-4\\"><div class=\\"timeline-content container position-relative\\"><div class=\\"timeline-item position-relative py-2 px-4 timeline-item-home timeline-item-left pr-5\\"><div id=\\"timeline-item-0\\" class=\\"timeline-item-anchor\\">&nbsp;</div> <div class=\\"timeline-item-bubble rounded-circle flag-icon-background flag-icon-squared flag-icon-2x timeline-item-bubble-left flag-icon-fr\\"></div> <div class=\\"timeline-item-content position-relative rounded p-2 p-lg-4 timeline-item-content-home\\"><fontawesomeicon-stub size=\\"4x\\" icon=\\"home\\" class=\\"timeline-item-content-icon mx-2\\"></fontawesomeicon-stub> <h1 class=\\"timeline-item-content-destination\\">
      Paris, France
    </h1> <span class=\\"badge timeline-item-content-status badge-light badge-pill\\">Home</span> <!----></div></div><div class=\\"timeline-item position-relative py-2 px-4 timeline-item-visited timeline-item-right pl-5\\"><div id=\\"timeline-item-1\\" class=\\"timeline-item-anchor\\">&nbsp;</div> <div class=\\"timeline-item-bubble rounded-circle flag-icon-background flag-icon-squared flag-icon-2x timeline-item-bubble-right flag-icon-pl\\"></div> <div class=\\"timeline-item-content position-relative rounded p-2 p-lg-4 timeline-item-content-visited\\"><fontawesomeicon-stub size=\\"4x\\" icon=\\"check-circle\\" class=\\"timeline-item-content-icon mx-2\\"></fontawesomeicon-stub> <h1 class=\\"timeline-item-content-destination\\">
      Warsaw, Poland
    </h1> <span class=\\"badge timeline-item-content-status badge-light badge-pill\\">Visited</span> <p class=\\"timeline-item-content-dates m-0 pl-2 pt-2\\"><span>From 1/1/2018</span> <span>&nbsp;to 6/30/2018</span></p></div></div><div class=\\"timeline-item position-relative py-2 px-4 timeline-item-visited timeline-item-left pr-5\\"><div id=\\"timeline-item-2\\" class=\\"timeline-item-anchor\\">&nbsp;</div> <div class=\\"timeline-item-bubble rounded-circle flag-icon-background flag-icon-squared flag-icon-2x timeline-item-bubble-left flag-icon-hr\\"></div> <div class=\\"timeline-item-content position-relative rounded p-2 p-lg-4 timeline-item-content-visited\\"><fontawesomeicon-stub size=\\"4x\\" icon=\\"check-circle\\" class=\\"timeline-item-content-icon mx-2\\"></fontawesomeicon-stub> <h1 class=\\"timeline-item-content-destination\\">
      Zagreb, Croatia
    </h1> <span class=\\"badge timeline-item-content-status badge-light badge-pill\\">Visited</span> <p class=\\"timeline-item-content-dates m-0 pl-2 pt-2\\"><span>From 7/1/2018</span> <span>&nbsp;to 12/31/2018</span></p></div></div><div class=\\"timeline-item position-relative py-2 px-4 timeline-item-visited timeline-item-right pl-5\\"><div id=\\"timeline-item-3\\" class=\\"timeline-item-anchor\\">&nbsp;</div> <div class=\\"timeline-item-bubble rounded-circle flag-icon-background flag-icon-squared flag-icon-2x timeline-item-bubble-right flag-icon-fr\\"></div> <div class=\\"timeline-item-content position-relative rounded p-2 p-lg-4 timeline-item-content-visited\\"><fontawesomeicon-stub size=\\"4x\\" icon=\\"check-circle\\" class=\\"timeline-item-content-icon mx-2\\"></fontawesomeicon-stub> <h1 class=\\"timeline-item-content-destination\\">
      Paris, France
    </h1> <span class=\\"badge timeline-item-content-status badge-light badge-pill\\">Visited</span> <p class=\\"timeline-item-content-dates m-0 pl-2 pt-2\\"><span>From 1/1/2019</span> <span>&nbsp;to 1/1/2019</span></p></div></div><div class=\\"timeline-item position-relative py-2 px-4 timeline-item-current timeline-item-left pr-5\\"><div id=\\"timeline-item-4\\" class=\\"timeline-item-anchor\\">&nbsp;</div> <div class=\\"timeline-item-bubble rounded-circle flag-icon-background flag-icon-squared flag-icon-2x timeline-item-bubble-left flag-icon-is\\"></div> <div class=\\"timeline-item-content position-relative rounded p-2 p-lg-4 timeline-item-content-current\\"><fontawesomeicon-stub size=\\"4x\\" icon=\\"map-marker\\" class=\\"timeline-item-content-icon mx-2\\"></fontawesomeicon-stub> <h1 class=\\"timeline-item-content-destination\\">
      Reykjav√≠k, Iceland
    </h1> <span class=\\"badge timeline-item-content-status badge-light badge-pill\\">Current destination</span> <p class=\\"timeline-item-content-dates m-0 pl-2 pt-2\\"><span>From 1/2/2019</span> <span>&nbsp;to 12/31/2098</span></p></div></div><div class=\\"timeline-item position-relative py-2 px-4 timeline-item-next timeline-item-right pl-5\\"><div id=\\"timeline-item-5\\" class=\\"timeline-item-anchor\\">&nbsp;</div> <div class=\\"timeline-item-bubble rounded-circle flag-icon-background flag-icon-squared flag-icon-2x timeline-item-bubble-right flag-icon-fr\\"></div> <div class=\\"timeline-item-content position-relative rounded p-2 p-lg-4 timeline-item-content-next\\"><fontawesomeicon-stub size=\\"4x\\" icon=\\"bookmark\\" class=\\"timeline-item-content-icon mx-2\\"></fontawesomeicon-stub> <h1 class=\\"timeline-item-content-destination\\">
      Paris, France
    </h1> <span class=\\"badge timeline-item-content-status badge-light badge-pill\\">Next destination</span> <p class=\\"timeline-item-content-dates m-0 pl-2 pt-2\\"><span>From 1/1/2099</span> <span>&nbsp;to 1/1/2099</span></p></div></div><div class=\\"timeline-item position-relative py-2 px-4 timeline-item-next timeline-item-left pr-5\\"><div id=\\"timeline-item-6\\" class=\\"timeline-item-anchor\\">&nbsp;</div> <div class=\\"timeline-item-bubble rounded-circle flag-icon-background flag-icon-squared flag-icon-2x timeline-item-bubble-left flag-icon-gr\\"></div> <div class=\\"timeline-item-content position-relative rounded p-2 p-lg-4 timeline-item-content-next\\"><fontawesomeicon-stub size=\\"4x\\" icon=\\"bookmark\\" class=\\"timeline-item-content-icon mx-2\\"></fontawesomeicon-stub> <h1 class=\\"timeline-item-content-destination\\">
      Athens, Greece
    </h1> <span class=\\"badge timeline-item-content-status badge-light badge-pill\\">Next destination</span> <p class=\\"timeline-item-content-dates m-0 pl-2 pt-2\\"><span>From 1/2/2099</span> <!----></p></div></div></div></div>"
`;
